pkg/webhook/preflight/nutanix/clients.go

```diff
@@ -83,126 +197,202 @@ func newClient(
-		return nil, err
-	}
-	return resp, nil
+) (
+	*clustermgmtv4.ListClustersApiResponse,
+	error,
+) {
+	return callWithContext(ctx, func() (*clustermgmtv4.ListClustersApiResponse, error) {
+		return c.ListClustersFunc(
+			page_,
+			limit_,
+			filter_,
+			orderby_,
+			apply_,
+			select_,
+			args...,
+		)
+	})
 }
 
 func (c *clientWrapper) ListStorageContainers(
+	ctx context.Context,
 	page_ *int,
 	limit_ *int,
 	filter_ *string,
 	orderby_ *string,
 	select_ *string,
 	args ...map[string]interface{},
-) (*clustermgmtv4.ListStorageContainersApiResponse, error) {
-	resp, err := c.v4client.StorageContainerAPI.ListStorageContainers(
-		page_,
-		limit_,
-		filter_,
-		orderby_,
-		select_,
-		args...,
-	)
-	if err != nil {
-		return nil, err
-	}
-	return resp, nil
+) (
+	*clustermgmtv4.ListStorageContainersApiResponse,
+	error,
+) {
+	return callWithContext(ctx, func() (*clustermgmtv4.ListStorageContainersApiResponse, error) {
+		return c.ListStorageContainersFunc(
+			page_,
+			limit_,
+			filter_,
+			orderby_,
+			select_,
+			args...,
+		)
+	})
 }
 
-func (c *clientWrapper) GetSubnetById(id *string) (*netv4.GetSubnetApiResponse, error) {
-	resp, err := c.v4client.SubnetsApiInstance.GetSubnetById(id)
-	if err != nil {
-		return nil, err
-	}
-	return resp, nil
+func (c *clientWrapper) GetSubnetById(
+	ctx context.Context,
+	uuid *string,
+	args ...map[string]interface{},
+) (
+	*netv4.GetSubnetApiResponse,
+	error,
+) {
+	return callWithContext(ctx, func() (*netv4.GetSubnetApiResponse, error) {
+		return c.GetSubnetByIdFunc(
+			uuid,
+			args...,
+		)
+	})
 }
 
 func (c *clientWrapper) ListSubnets(
+	ctx context.Context,
 	page_ *int,
 	limit_ *int,
 	filter_ *string,
 	orderby_ *string,
 	expand_ *string,
 	select_ *string,
 	args ...map[string]interface{},
-) (*netv4.ListSubnetsApiResponse, error) {
-	resp, err := c.v4client.SubnetsApiInstance.ListSubnets(
-		page_,
-		limit_,
-		filter_,
-		orderby_,
-		expand_,
-		select_,
-		args...,
-	)
-	if err != nil {
-		return nil, err
+) (
+	*netv4.ListSubnetsApiResponse,
+	error,
+) {
+	return callWithContext(ctx, func() (*netv4.ListSubnetsApiResponse, error) {
+		return c.ListSubnetsFunc(
+			page_,
+			limit_,
+			filter_,
+			orderby_,
+			expand_,
+			select_,
+			args...,
+		)
+	})
+}
+
+// callWithContext is a helper function that immediately responds to context cancellation,
+// while calling a long-running, non-preemptible function. The long-running function always
+// runs to completion, but its result is only returned if the context is not cancelled.
+func callWithContext[T any](ctx context.Context, f func() (T, error)) (resp T, err error) {
+	respCh := make(chan T)
+	errCh := make(chan error)
+
+	go func() {
+		resp, err := f()
+		select {
+		case <-ctx.Done():
+			// Context was cancelled before function returned. We assume no one wants the result anymore.
+		default:
+			if err != nil {
+				errCh <- err
+			}
+			respCh <- resp
+		}
+		close(respCh)
+		close(errCh)
+	}()
```

Comment:
If you use a buffered channel, you can simplify to
```go
go func() {
        v, e := f()
        ch <- result[T]{value: v, err: e}
}()
```

Reply:
It's simpler, but I think it's harder to read/understand.

How do you know what happens to the result that's sent to to `ch`? That is, how can you be sure the channel will be garbage collected? You have to know that the result is sent to the buffer, and that if the `select` statement already terminated, then the channel will never be read, so once the enclosing function goes out of scope, the channel will be garbage-collected.

I guess I could explain this in a comment, though.

Reply:
You convinced me :smile:. Thanks!

Done in https://github.com/nutanix-cloud-native/cluster-api-runtime-extensions-nutanix/pull/1234/commits/c74b9a615e300de61a6e0259a065d1a5fdfb0fae